inc/appliance.class.php

```diff
@@ -0,0 +1,626 @@
+         'WHERE'  => ['id' => $sub_query],
+         'ORDER'  => 'name'
+      ];
+
+      $result = $DB->request($query);
+
+      $values = [0 => Dropdown::EMPTY_VALUE];
+
+      while ($data =$result->next()) {
+         $values[$data['id']] = $data['name'];
+      }
+      $rand     = mt_rand();
+      $out      = Dropdown::showFromArray(
+         '_appliancetype',
+         $values, [
+            'width'   => '30%',
+            'rand'    => $rand,
+            'display' => false
+         ]
+      );
+      $field_id = Html::cleanId("dropdown__appliancetype$rand");
+
+      $params   = [
+         'appliancetype' => '__VALUE__',
+         'entity'        => $p['entity'],
+         'rand'          => $rand,
+         'myname'        => $p['name'],
+         'used'          => $p['used']
+      ];
+
+      $out .= Ajax::updateItemOnSelectEvent(
+         $field_id, "show_".$p['name'].$rand,
+         $CFG_GLPI["root_doc"]."/ajax/dropdownTypeAppliances.php",
+         $params,
+         false
+      );
+      $out .= "<span id='show_".$p['name']."$rand'>";
+      $out .= "</span>\n";
+
+      $params['appliancetype'] = 0;
+      $out .= Ajax::updateItem(
+         "show_".$p['name'].$rand,
+         $CFG_GLPI["root_doc"]. "/ajax/dropdownTypeAppliances.php",
+         $params,
+         false
+      );
+      if ($p['display']) {
+         echo $out;
+         return $rand;
+      }
+      return $out;
+   }
+
+
+   /**
+    * Type than could be linked to a Appliance
+    *
+    * @param boolean $all All types  or only allowed ones (defauts to false)
+    *
+    * @return array of types
+   **/
+   static function getTypes($all = false) {
+      global $CFG_GLPI;
+
+      $types = $CFG_GLPI['appliance_types'];
+
+      if ($all) {
+         return $types;
+      }
+
+      // Only allowed types
+      foreach ($types as $key => $type) {
+         if (!($item = getItemForItemtype($type))) {
+            continue;
+         }
+
+         if (!$item->canView()) {
+            unset($types[$key]);
+         }
+      }
+      return $types;
+   }
+
+
+   function getSpecificMassiveActions($checkitem = null) {
+
+      $isadmin = static::canUpdate();
+      $actions = parent::getSpecificMassiveActions($checkitem);
+
+      if ($_SESSION['glpiactiveprofile']['interface'] == 'central') {
+         if ($isadmin) {
+            $actions['Appliance'.MassiveAction::CLASS_ACTION_SEPARATOR.'install'] = _x('button', 'Associate');
+            $actions['Appliance'.MassiveAction::CLASS_ACTION_SEPARATOR.'uninstall'] = _x('button', 'Dissociate');
+
+            if (Session::haveRight('transfer', READ) && Session::isMultiEntitiesMode()) {
+               $actions['Appliance'.MassiveAction::CLASS_ACTION_SEPARATOR.'transfer'] = __('Transfer');
+            }
+         }
+      }
+      return $actions;
+   }
+
+
+   static function showMassiveActionsSubForm(MassiveAction $ma) {
+
+      switch ($ma->getAction()) {
+         case 'add_item':
```

Comment:
This one is not accessible for now. I guess you should add a `getMassiveActionsForItemtype` method in `Appliance` and call it from `getSpecificMassiveActions` of itemtypes that should propose this action.

Reply:
See https://github.com/glpi-project/glpi/pull/7052/commits/b59351c07590667d23adaf8e1db02aa0fdfa2132